File: selection.cpp
#include <iostream>
using std::cout;
using std::endl;

#include <iomanip>
using std::setw;

bool valid_l(int l, int n) { return l >= 0 and l < n; }
bool valid_m(int m, int l) { return m >= -l and m <= l; }

int main()
{
   const int Nmax = 20;
   for (int n2 = 1; n2 <= Nmax; ++n2)
      for (int n1 = 1; n1 < n2; ++n1)
      {
         int num = 0;
         for (int l2 = 0; l2 < n2; ++l2)
            for (int m2 = -l2; m2 <= l2; ++m2)
            {
               // the l1 = l2-1 transition
               if (valid_l(l2-1,n1))
               {
                  const int l1 = l2-1;
                  if (valid_m(m2-1,l1)) ++num;
                  if (valid_m(m2,l1)) ++num;
                  if (valid_m(m2+1,l1)) ++num;
               }
               // the l1 = l2+1 transition
               if (valid_l(l2+1,n1))
               {
                  const int l1 = l2+1;
                  if (valid_m(m2-1,l1)) ++num;
                  if (valid_m(m2,l1)) ++num;
                  if (valid_m(m2+1,l1)) ++num;
               }
            }
         const double dE = 13.6*(1.0/n1/n1 - 1.0/n2/n2);
         const double lambda = 1240.0/dE;
         cout << setw(2) << n2 << "->" << n1 << (lambda >= 380.0 and lambda <= 750.0 ? 'v' : ' ')
         << setw(10) << num << setw(16) << dE << setw(16) << lambda << endl;
      }
   return 0;
}

File: selection.rs
fn valid_l(l:i64, n:i64) -> bool { l >= 0 && l < n }
fn valid_m(m:i64, l:i64) -> bool { m >= -l && m <= l }

fn main() {
   let nmax = 20;
   for n2 in 1..=nmax {
      for n1 in 1..n2 {
         let mut num = 0;
         for l2 in 0..n2 {
            for m2 in -l2..=l2 {
               // the l1 = l2-1 transition
               if valid_l(l2-1,n1) {
                  let l1 = l2-1;
                  if valid_m(m2-1,l1) { num+=1; }
                  if valid_m(m2,l1) { num+=1; }
                  if valid_m(m2+1,l1) { num+=1; }
               }
               // the l1 = l2+1 transition
               if valid_l(l2+1,n1) {
                  let l1 = l2+1;
                  if valid_m(m2-1,l1) { num+=1; }
                  if valid_m(m2,l1) { num+=1; }
                  if valid_m(m2+1,l1) { num+=1; }
               }
            }
         }
         let term1 = 1.0/(n1 as f64)/(n1 as f64);
         let term2 = 1.0/(n2 as f64)/(n2 as f64);
         let de = 13.6*(term1 - term2);
         let lambda = 1240.0/de;
         println!("{:2} -> {:2}{:1} {:10}{:20.10e}{:20.10e}", n2, n1,
                  if lambda >= 380.0 && lambda <= 750.0 
                  { 'v' } else {' '}, num, de, lambda);
      }
   }
}

